infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/wsplugins/OpenshiftPluginBrokerManager.java

```diff
@@ -0,0 +1,118 @@
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.WorkspaceVolumesStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.RuntimeEventsPublisher;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.UnrecoverablePodEventListenerFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.BrokersResult;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.KubernetesPluginsToolingValidator;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.PluginBrokerManager;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.BrokerEnvironmentFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.DeployBroker;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.ListenBrokerEvents;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.PrepareStorage;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.WaitBrokerResult;
+import org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProject;
+import org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProjectFactory;
+import org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner;
+
+/**
+ * Deploys and runs Che plugin metadata broker. If Openshift 4 is used, then mounts trusted CA
+ * bundle into the broker container. Waits till the broker finishes its work and returns resolved
+ * workspace tooling or error if any.
+ */
+public class OpenshiftPluginBrokerManager<E extends KubernetesEnvironment>
+    extends PluginBrokerManager<E> {
+
+  private final OpenShiftProjectFactory factory;
+  private final Openshift4TrustedCAProvisioner trustedCAProvisioner;
+
+  @Inject
+  public OpenshiftPluginBrokerManager(
+      OpenShiftProjectFactory factory,
+      EventService eventService,
+      KubernetesPluginsToolingValidator pluginsValidator,
+      KubernetesEnvironmentProvisioner<E> environmentProvisioner,
+      WorkspaceVolumesStrategy volumesStrategy,
+      BrokerEnvironmentFactory<E> brokerEnvironmentFactory,
+      UnrecoverablePodEventListenerFactory unrecoverablePodEventListenerFactory,
+      @Named("che.workspace.plugin_broker.wait_timeout_min") int pluginBrokerWaitingTimeout,
+      RuntimeEventsPublisher runtimeEventsPublisher,
+      Tracer tracer,
+      Openshift4TrustedCAProvisioner trustedCAProvisioner) {
+    super(
+        factory,
+        eventService,
+        pluginsValidator,
+        environmentProvisioner,
+        volumesStrategy,
+        brokerEnvironmentFactory,
+        unrecoverablePodEventListenerFactory,
+        pluginBrokerWaitingTimeout,
+        runtimeEventsPublisher,
+        tracer);
+    this.trustedCAProvisioner = trustedCAProvisioner;
+    this.factory = factory;
+  }
+
+  @Beta
+  @Traced
+  public List<ChePlugin> getTooling(
```

Comment:
Is not it pretty the same as Kubernetes implementation? Can't we avoid that duplication?

Reply:
I will try and test

Reply:
@sleshchenko I have tried your proposal and it works too. So I agree that it is more optimal way of resolving the problem and I've added corresponding changes into this PR. Please take a look again.